python/ray/experimental/data/dataset.py

```diff
@@ -195,6 +196,117 @@ def repartition(self, num_blocks: int) -> "Dataset[T]":
+
+        # Split self._blocks into n chunks. the first {num_large_chunks} chunks
+        # have {small_chunk_size + 1} blocks, and remaining
+        # chunks have {small_chunk_size} blocks.
+        small_chunk_size = len(self._blocks) // n
+        num_large_chunks = len(self._blocks) - small_chunk_size * n
+
+        datasets = []
+        block_refs = list(self._blocks)
+
+        if locality_hints is None:
+            cur = 0
+            for i in range(n):
+                end = cur + small_chunk_size
+                if i < num_large_chunks:
+                    end = end + 1
+                datasets.append(Dataset(block_refs[cur:end]))
+                cur = end
+            return datasets
+
+        # If the locality_hints is set, we use a greedy algorithm
+        # to co-locate the blocks with the actors based on block
+        # and actor's node_id:
+        #
+        # 1. We get the node_id where the block resides in and store it in
+        # a map {node_id: [blocks]}. Note that each block could
+        # be located on multiple nodes. For simplicity, we use the first
+        # node_id so that there is one to one mapping from node_id to block
+        # in the map.
+        #
+        # 2. We do two round of allocations. In the first round, we allocate
+        # blocks for each actor based by sustract the blocks from the map
+        # we build from previous step.
+        #
+        # 3. Some actors might have more blocks than needed while the
+        # others don't have enough blocks. We do second round of allocation
+        # for the reminder blocks so that every actor have sufficient blocks.
+
+        # wait might affect the throughput due to stragglers?
+        ray.wait(block_refs)
+
+        # build node_id to block_refs map
+        block_ref_locations = ray.experimental.get_object_locations(block_refs)
+        block_refs_by_node_id = {}
+        for block_ref in block_refs:
+            node_ids = block_ref_locations.get(block_ref, {}).get(
+                "node_ids", [])
+            node_id = node_ids[0] if node_ids else None
+            if node_id not in block_refs_by_node_id:
+                block_refs_by_node_id[node_id] = []
+            block_refs_by_node_id[node_id].append(block_ref)
+
+        # find the node_ids for each actors and the num of blocks
+        # needed for each actor
+        actors_state = ray.state.actors()
+        actor_node_ids = [
+            actors_state.get(actor._actor_id.hex(), {}).get("Address",
+                                                            {}).get("NodeID")
+            for actor in locality_hints
+        ]
+        num_blocks_by_actor = [small_chunk_size + 1] * num_large_chunks + [
+            small_chunk_size
+        ] * (n - num_large_chunks)
+
+        # first round of allocation by substracting the blocks from
+        # block_refs_by_node_id map based on each actor's node_id.
+        block_refs_by_actor = []
+        for i in range(n):
```

Comment:
Shouldn't this repeat until there are no local assignable blocks? How about this algorithm instead:

```
actor_to_assigned_blocks = collections.defaultdict(list)

# First assign as many blocks as we can that are local to the actor.
any_assigned = True:
while any_assigned:
  any_assigned = False
  for actor, location in actor_locations:
      local_block = allocate_local_block(blocks_by_node_id, location)
      if local_block:
          actor_to_assigned_blocks[actor].append(local_block)
          any_assigned = True

# Then fill in random blocks to try to balance the number of blocks per actor.
```

Reply:
We could also draw inspiration from https://en.wikipedia.org/wiki/Maximum_cardinality_matching but this is probably overkill for now--- especially since it's highly likely each block is only present on  one node.

Reply:
yeah i think that's equivalent... refactor the code a bit for better readability. let me know if it looks better.